File: Items.py
from GlobalVariables import entities, items, chunks, listOfBlockNames, blockSize, gravity, dictOfBlockBreakingStuff
from Worldgen import findBlock, getChunkCoord, getBlockCoord, smallScaleBlockUpdates
from Controls import mouse
from Entities import ItemEntity



# basic item, default values to make it exist in player inventory properly
class Item():
    def __init__(self, name = "air"):
        self.name = name
        self.slotId = 0
        self.tooltip = "tooltip unavailable"
        
    
    # drop/throw this item out as an entity
    def drop(self, x, y, z, xv, yv, zv, player = None, count = 1):

        droppedItem = ItemEntity(self, count, x, y, z, xv, yv, zv)

        entities.append(droppedItem)

        if player != None:
            print("doing stuff to player inventory after dropping something")
            if player.hotbar[self.slotId]["count"] == 1:
                player.hotbar[self.slotId]["contents"] = "empty"
            elif player.hotbar[self.slotId]["count"] >= 2:
                player.hotbar[self.slotId]["count"] -= 1

# an item that can be placed, like wood or something
class PlaceableItem(Item):
    def __init__(self, name, tooltip = ""):
        super().__init__(name)
        self.tooltip = self.name
        if tooltip != "":
            self.tooltip = tooltip
        self.slotId = 0
        self.itemType = "PlaceableItem"
        self.stackable = True
        self.placedBlock = {
            "type": self.name,
            "render": False,
            "alphaValue": 0,
            "hardness": 0,
            "effectiveTool": "none"
        }
        self.placedBlock["hardness"] = dictOfBlockBreakingStuff[self.name]["hardness"]
        self.placedBlock["effectiveTool"] = dictOfBlockBreakingStuff[self.name]["effectiveTool"]
    def placeItem(self, player):

        
        blockType = mouse.hoveredBlock["block"]["type"]
        if blockType == "air" or blockType == "water":
            count = player.hotbar[self.slotId]["count"]
            if count > 1:
                player.hotbar[self.slotId]["count"] -= 1
                
            elif count == 1:
                player.hotbar[self.slotId]["count"] = 0
                player.hotbar[self.slotId]["contents"] = "empty"
                
            elif count <= 0:
                print("what the heck, how did you do that??")

            chunkCoord = mouse.hoveredBlock["chunkCoord"]
            blockCoord = mouse.hoveredBlock["blockCoord"]
            

            chunks[chunkCoord]["data"][blockCoord] = self.placedBlock

            smallScaleBlockUpdates(chunkCoord, blockCoord)



    def RMBAction(self, player):
        pass
        

    def RMBPressedAction(self, player):
        self.placeItem(player)

class ToolItem(Item):
    def __init__(self, name, toolData = {"attack": 1, "knockback": 1,
                                         "breakingPower": 1, "breakingSpeed": 1,
                                         "breakingType": "none"},
                                         tooltip = ""):
        super().__init__(name)
        self.slotId = 0
        self.itemType = "ToolItem"
        self.tooltip = self.name
        if tooltip != "":
            self.tooltip = tooltip
        self.stackable = False
        
        self.attack = toolData["attack"]
        self.knockback = toolData["knockback"]
        self.breakingPower = toolData["breakingPower"]
        self.breakingSpeed = toolData["breakingSpeed"]
        self.breakingType = toolData["breakingType"]

        self.durability = 100 # durability will exist when the game actually functions
    
    def RMBPressedAction(self, player):
        pass

    def RMBAction(self, player):
        pass


# adding items to the game
def addItem(name = "air", itemType = "none", toolData = {}):

    if itemType == "placeable":
        item = PlaceableItem(name)
    
    if itemType == "tool":
        item = ToolItem(name, toolData)

    items[name] = item

def makeItemsExist():
    for itemName in listOfBlockNames:
        addItem(itemName, "placeable")
    addItem("stone pickaxe", "tool",
            {"attack": 3, "knockback": 1, "breakingPower": 3,
            "breakingSpeed": 20, "breakingType": "pickaxe"})
    
    addItem("stone axe", "tool",
            {"attack": 7, "knockback": 1, "breakingPower": 3,
            "breakingSpeed": 20, "breakingType": "axe"})
    
    addItem("stone shovel", "tool",
            {"attack": 2, "knockback": 1, "breakingPower": 1,
            "breakingSpeed": 20, "breakingType": "shovel"})







print("items initialized")

File: Items.mjs


import {
    entities, items, chunks, listOfBlockNames,
    listOfIntermediateItems, blockSize, gravity,
    dictOfBlockBreakingStuff,
    showLoadingProgress, mouse, maxStackSize
} from "./GlobalVariables.mjs";
showLoadingProgress("loading Items.mjs")

//import { mouse } from "./Controls.mjs";
import { ItemEntity } from "./Entities.mjs";
import { smallScaleBlockUpdates } from "./Worldgen.mjs";


class Item {
    constructor(name) {


        this.name = name || "air";
        this.slotId = 0;
        this.tooltip = "tooltip unavailable";
        this.stackable = false;


        this.drop = function (x, y, z, xv, yv, zv, count = 1, dropSource = undefined, actualDropSource) {

            let droppedItem = new ItemEntity(x, y, z, xv, yv, zv, count, this);
            entities.push(droppedItem);


            if (dropSource === "mouse") {
                mouse.heldSlot.count = 0;
                mouse.heldSlot.contents = "empty";
            }

            if (dropSource === "player") {
                if (actualDropSource.hotbar[this.slotId]["count"] == 1) {
                    actualDropSource.hotbar[this.slotId]["contents"] = "empty";
                } else {
                    if (actualDropSource.hotbar[this.slotId].count >= 2) {
                        actualDropSource.hotbar[this.slotId]["count"] -= 1
                    };
                };
            };


        };

    };
};


export class PlaceableItem extends Item {
    constructor(placeableType, stackable) {
        super();

        this.name = placeableType || this.name;
        this.stackable = stackable || true;
        this.itemType = "PlaceableItem";
        this.tooltip = this.name;

        this.placedBlock = {
            "type": this.name,
            "render": false,
            "alphaValue": 1,
            "hardness": 0,
            "effectiveTool": "none",
            "dropsWithNoTool": false
        };

        this.placedBlock.hardness = dictOfBlockBreakingStuff[this.name].hardness;
        this.placedBlock.effectiveTool = dictOfBlockBreakingStuff[this.name].effectiveTool;
        this.placedBlock.dropsWithNoTool = dictOfBlockBreakingStuff[this.name].dropsWithNoTool;

        this.placeItem.bind(this);
        this.RMBAction.bind(this);
        this.RMBPressedAction.bind(this);

    };

    placeItem(player) {

        if (player.canReachSelectedBlock) {
            let blockType = mouse.hoveredBlock.type;

            if (blockType == "air" || blockType == "water") {
                let count = player.hotbar[this.slotId].count;

                if (count >= 1) {


                    if (count == 1) {
                        player.hotbar[this.slotId].count = 0;
                        player.hotbar[this.slotId].contents = "empty";
                    } else if (count > 1) {
                        player.hotbar[this.slotId].count -= 1;
                    }


                    let chunkCoord = mouse.hoveredBlock.chunkCoord;
                    let blockCoord = mouse.hoveredBlock.blockCoord;


                    chunks[chunkCoord].data[blockCoord] = this.placedBlock;


                    smallScaleBlockUpdates(chunkCoord, blockCoord);
                };
            };
        };

    };

    RMBAction = function (player) {

    };

    RMBPressedAction = function (player) {
        this.placeItem(player);
    };

};

export class ToolItem extends Item {
    constructor(name, toolData = {
        "attack": 1, "knockback": 1,
        "breakingPower": 1, "breakingSpeed": 1,
        "breakingType": "none"
    },
        tooltip, stackable) {
        super([name])
        this.slotId = 0
        this.itemType = "ToolItem"
        this.tooltip = tooltip || this.name;

        this.stackable = stackable || false;

        this.attack = toolData.attack
        this.knockback = toolData.knockback
        this.breakingPower = toolData.breakingPower
        this.breakingSpeed = toolData.breakingSpeed
        this.breakingType = toolData.breakingType

        this.durability = 100; // durabilty does nothing atm

        this.RMBPressedAction.bind(this);
        this.RMBAction.bind(this);

    };

    RMBPressedAction = function (player) {

    };


    RMBAction = function (player) {

    };
};

class IntermediateItem extends Item {
    constructor(name, stackable, slotId) {
        super();
        this.name = name;
        this.stackable = stackable || true;
        this.tooltip = this.name
        this.slotId = slotId || 0;
    };
}


function addItem(name = "air", itemType = "none", toolData = {}, stackable = false) {

    let item = undefined;
    if (itemType == "placeable") {
        item = new PlaceableItem(name, stackable = true);
    };
    if (itemType == "tool") {
        item = new ToolItem(name, toolData);
    };
    if (itemType == "intermediate") {
        item = new IntermediateItem(name, stackable);
    };

    if (item !== undefined) { items[name] = item; };
};

export function makeItemsExist() {
    listOfBlockNames.forEach(function (itemName) {
        addItem(itemName, "placeable", undefined, true);
    });
    listOfIntermediateItems.forEach(function (itemName) {
        addItem(itemName, "intermediate", undefined, true);
    });
    addItem("stone pickaxe", "tool",
        {
            "attack": 3, "knockback": 1, "breakingPower": 3,
            "breakingSpeed": 20, "breakingType": "pickaxe"
        })

    addItem("stone axe", "tool",
        {
            "attack": 7, "knockback": 1, "breakingPower": 3,
            "breakingSpeed": 20, "breakingType": "axe"
        })

    addItem("stone shovel", "tool",
        {
            "attack": 2, "knockback": 1, "breakingPower": 1,
            "breakingSpeed": 25, "breakingType": "shovel"
        })


    addItem("wood pickaxe", "tool",
        {
            "attack": 2, "knockback": 1, "breakingPower": 2,
            "breakingSpeed": 10, "breakingType": "pickaxe"
        })

    addItem("wood axe", "tool",
        {
            "attack": 6, "knockback": 1, "breakingPower": 2,
            "breakingSpeed": 10, "breakingType": "axe"
        })

    addItem("wood shovel", "tool",
        {
            "attack": 1, "knockback": 1, "breakingPower": 1,
            "breakingSpeed": 15, "breakingType": "shovel"
        })
};






showLoadingProgress("Items.mjs loaded");
